File: contextual-illustrator/scripts/generate_image.py
#!/usr/bin/env python3
"""Generate images via fal.ai or OpenRouter.

Supported models:
  - gpt-image-2   : OpenAI GPT-Image-2 via fal.ai (default; text-to-image and edit)
  - gemini-3-pro  : Gemini 3 Pro Image Preview (fal.ai or OpenRouter)
"""

import argparse
import base64
import json
import os
import sys
import urllib.request
import urllib.error


def find_env_file():
    """Search for .env file: skill root first, then cwd upward."""
    # 1. Check skill root directory (parent of scripts/)
    script_dir = os.path.dirname(os.path.abspath(__file__))
    skill_root = os.path.dirname(script_dir)
    skill_env = os.path.join(skill_root, ".env")
    if os.path.isfile(skill_env):
        return skill_env

    # 2. Fallback: search from cwd upward
    path = os.getcwd()
    while True:
        env_path = os.path.join(path, ".env")
        if os.path.isfile(env_path):
            return env_path
        parent = os.path.dirname(path)
        if parent == path:
            return None
        path = parent


def load_env_keys() -> dict[str, str]:
    """Load API keys from .env file. Returns dict of found keys."""
    keys = {}

    # Check environment variables first
    for var in ("FAL_KEY", "OPENROUTER_API_KEY"):
        if os.environ.get(var):
            keys[var] = os.environ[var]

    env_path = find_env_file()
    if not env_path:
        return keys

    with open(env_path) as f:
        for line in f:
            line = line.strip()
            if not line or line.startswith("#"):
                continue
            if "=" in line:
                key, _, value = line.partition("=")
                key = key.strip()
                value = value.strip().strip('"').strip("'")
                if key in ("FAL_KEY", "OPENROUTER_API_KEY") and value:
                    keys[key] = value

    return keys


def detect_backend(keys: dict[str, str]) -> str:
    """Auto-detect backend based on available keys. Prefer fal.ai."""
    if "FAL_KEY" in keys:
        return "fal"
    if "OPENROUTER_API_KEY" in keys:
        return "openrouter"
    return "none"


# Aspect-ratio → fal.ai image_size preset for GPT-Image-2.
# Falls back to `landscape_4_3` for unknown ratios.
_GPT2_ASPECT_TO_SIZE = {
    "1:1": "square_hd",
    "16:9": "landscape_16_9",
    "4:3": "landscape_4_3",
    "3:2": "landscape_4_3",
    "5:4": "square_hd",
    "4:5": "portrait_4_3",
    "3:4": "portrait_4_3",
    "2:3": "portrait_4_3",
    "9:16": "portrait_16_9",
    "21:9": "landscape_16_9",
    "auto": "landscape_4_3",
}


def _resolve_gpt2_image_size(image_size: str | None, aspect_ratio: str):
    """Resolve GPT-Image-2 image_size param.

    Accepts:
      - preset name string (e.g. 'landscape_4_3') — passed through
      - explicit 'WxH' (e.g. '1024x1024') — converted to {width, height}
      - None — derived from --aspect-ratio
    """
    if image_size:
        if "x" in image_size.lower():
            try:
                w, h = image_size.lower().split("x", 1)
                return {"width": int(w), "height": int(h)}
            except ValueError:
                print(f"ERROR: invalid --image-size '{image_size}', expected WxH or preset name", file=sys.stderr)
                sys.exit(1)
        return image_size
    return _GPT2_ASPECT_TO_SIZE.get(aspect_ratio, "landscape_4_3")


# ---------------------------------------------------------------------------
# OpenRouter backend (Gemini 3 Pro only; stdlib, no extra deps)
# ---------------------------------------------------------------------------

def generate_via_openrouter(
    api_key: str,
    prompt: str,
    num_images: int = 1,
    aspect_ratio: str = "1:1",
    resolution: str = "1K",
    output_format: str = "png",
    output_dir: str = ".",
    filename_prefix: str = "illustration",
    **_kwargs,
) -> dict:
    """Generate images via OpenRouter Chat Completions API (Gemini 3 Pro)."""

    image_instruction = ""
    if num_images > 1:
        image_instruction = f" Generate {num_images} different variations."

    messages = [
        {
            "role": "user",
            "content": f"{prompt}{image_instruction}",
        }
    ]

    # image_config: aspect_ratio + image_size (1K/2K/4K)
    image_config = {}
    if aspect_ratio and aspect_ratio != "auto":
        image_config["aspect_ratio"] = aspect_ratio
    if resolution:
        image_config["image_size"] = resolution

    body = {
        "model": "google/gemini-3-pro-image-preview",
        "messages": messages,
        "modalities": ["image", "text"],
    }
    if image_config:
        body["image_config"] = image_config

    payload = json.dumps(body).encode()

    req = urllib.request.Request(
        "https://openrouter.ai/api/v1/chat/completions",
        data=payload,
        headers={
            "Content-Type": "application/json",
            "Authorization": f"Bearer {api_key}",
        },
        method="POST",
    )

    print("[openrouter] Sending request...", file=sys.stderr)
    try:
        with urllib.request.urlopen(req, timeout=120) as resp:
            result = json.loads(resp.read().decode())
    except urllib.error.HTTPError as e:
        body = e.read().decode() if e.fp else ""
        print(f"ERROR: OpenRouter API returned {e.code}: {body}", file=sys.stderr)
        sys.exit(1)

    # Extract images from response
    os.makedirs(output_dir, exist_ok=True)
    saved_files = []
    urls = []

    message = result.get("choices", [{}])[0].get("message", {})

    # OpenRouter returns images in parts (multimodal content) or in message.images
    images_data = []

    # Format 1: message.images[]
    if message.get("images"):
        for img in message["images"]:
            url = img.get("image_url", {}).get("url", "")
            if url:
                images_data.append(url)

    # Format 2: message.content as array with image parts
    elif isinstance(message.get("content"), list):
        for part in message["content"]:
            if part.get("type") == "image_url":
                url = part.get("image_url", {}).get("url", "")
                if url:
                    images_data.append(url)

    for i, image_url in enumerate(images_data):
        suffix = f"_{i+1}" if len(images_data) > 1 else ""
        filename = f"{filename_prefix}{suffix}.{output_format}"
        filepath = os.path.join(output_dir, filename)

        if image_url.startswith("data:"):
            # base64 data URL: data:image/png;base64,xxxxx
            header, _, b64data = image_url.partition(",")
            img_bytes = base64.b64decode(b64data)
            with open(filepath, "wb") as f:
                f.write(img_bytes)
        else:
            # Remote URL
            urllib.request.urlretrieve(image_url, filepath)
            urls.append(image_url)

        saved_files.append(filepath)
        print(f"[openrouter] Saved: {filepath}", file=sys.stderr)

    # Extract text description if present
    description = ""
    if isinstance(message.get("content"), str):
        description = message["content"]
    elif isinstance(message.get("content"), list):
        text_parts = [p.get("text", "") for p in message["content"] if p.get("type") == "text"]
        description = "\n".join(text_parts)

    return {
        "backend": "openrouter",
        "model": "gemini-3-pro",
        "files": saved_files,
        "urls": urls,
        "description": description,
    }


# ---------------------------------------------------------------------------
# fal.ai backend — Gemini 3 Pro (requires fal-client)
# ---------------------------------------------------------------------------

def _import_fal_client():
    try:
        import fal_client
        return fal_client
    except ImportError:
        print("ERROR: fal-client not installed. Run: cd <skill_dir> && uv venv && uv pip install fal-client", file=sys.stderr)
        sys.exit(1)


def _download_fal_images(result: dict, output_format: str, num_images: int,
                        output_dir: str, filename_prefix: str, backend_tag: str):
    os.makedirs(output_dir, exist_ok=True)
    saved_files = []
    urls = []
    for i, img in enumerate(result.get("images", [])):
        url = img["url"]
        urls.append(url)
        suffix = f"_{i+1}" if num_images > 1 else ""
        filename = f"{filename_prefix}{suffix}.{output_format}"
        filepath = os.path.join(output_dir, filename)
        urllib.request.urlretrieve(url, filepath)
        saved_files.append(filepath)
        print(f"[{backend_tag}] Saved: {filepath}", file=sys.stderr)
    return saved_files, urls


def generate_via_fal_gemini(
    api_key: str,
    prompt: str,
    num_images: int = 1,
    aspect_ratio: str = "1:1",
    output_format: str = "png",
    safety_tolerance: str = "4",
    resolution: str = "1K",
    seed: int | None = None,
    output_dir: str = ".",
    filename_prefix: str = "illustration",
    **_kwargs,
) -> dict:
    """Generate images via fal.ai Gemini 3 Pro Image Preview."""
    fal_client = _import_fal_client()
    os.environ["FAL_KEY"] = api_key

    arguments = {
        "prompt": prompt,
        "num_images": num_images,
        "aspect_ratio": aspect_ratio,
        "output_format": output_format,
        "safety_tolerance": safety_tolerance,
        "resolution": resolution,
    }
    if seed is not None:
        arguments["seed"] = seed

    def on_queue_update(update):
        if isinstance(update, fal_client.InProgress):
            for log in update.logs:
                print(f"  [log] {log['message']}", file=sys.stderr)

    print("[fal:gemini-3-pro] Sending request...", file=sys.stderr)
    result = fal_client.subscribe(
        "fal-ai/gemini-3-pro-image-preview",
        arguments=arguments,
        with_logs=True,
        on_queue_update=on_queue_update,
    )

    saved_files, urls = _download_fal_images(
        result, output_format, num_images, output_dir, filename_prefix, "fal:gemini-3-pro"
    )

    return {
        "backend": "fal",
        "model": "gemini-3-pro",
        "files": saved_files,
        "urls": urls,
        "description": result.get("description", ""),
    }


# ---------------------------------------------------------------------------
# fal.ai backend — OpenAI GPT-Image-2 (text-to-image and edit)
# ---------------------------------------------------------------------------

def generate_via_fal_gpt_image_2(
    api_key: str,
    prompt: str,
    num_images: int = 1,
    aspect_ratio: str = "1:1",
    image_size: str | None = None,
    quality: str = "high",
    output_format: str = "png",
    image_urls: list[str] | None = None,
    mask_url: str | None = None,
    output_dir: str = ".",
    filename_prefix: str = "illustration",
    **_kwargs,
) -> dict:
    """Generate images via fal.ai openai/gpt-image-2 (text-to-image or edit)."""
    fal_client = _import_fal_client()
    os.environ["FAL_KEY"] = api_key

    is_edit = bool(image_urls)
    endpoint = "openai/gpt-image-2/edit" if is_edit else "openai/gpt-image-2"

    arguments = {
        "prompt": prompt,
        "num_images": num_images,
        "quality": quality,
        "output_format": output_format,
    }

    if is_edit:
        arguments["image_urls"] = image_urls
        if mask_url:
            arguments["mask_url"] = mask_url
        # Edit endpoint defaults image_size to 'auto' — only set when user opts in
        if image_size:
            arguments["image_size"] = _resolve_gpt2_image_size(image_size, aspect_ratio)
    else:
        arguments["image_size"] = _resolve_gpt2_image_size(image_size, aspect_ratio)

    def on_queue_update(update):
        if isinstance(update, fal_client.InProgress):
            for log in update.logs:
                print(f"  [log] {log['message']}", file=sys.stderr)

    tag = "fal:gpt-image-2/edit" if is_edit else "fal:gpt-image-2"
    print(f"[{tag}] Sending request...", file=sys.stderr)
    result = fal_client.subscribe(
        endpoint,
        arguments=arguments,
        with_logs=True,
        on_queue_update=on_queue_update,
    )

    saved_files, urls = _download_fal_images(
        result, output_format, num_images, output_dir, filename_prefix, tag
    )

    return {
        "backend": "fal",
        "model": "gpt-image-2",
        "endpoint": endpoint,
        "files": saved_files,
        "urls": urls,
        "description": result.get("description", ""),
    }


# ---------------------------------------------------------------------------
# Main
# ---------------------------------------------------------------------------

def main():
    parser = argparse.ArgumentParser(
        description="Generate images (GPT-Image-2 default, or Gemini 3 Pro; auto-routes between fal.ai and OpenRouter)"
    )
    parser.add_argument("prompt", help="Text prompt for image generation")
    parser.add_argument("--model", default="gpt-image-2",
                        choices=["gemini-3-pro", "gpt-image-2"],
                        help="Which image model to use (default: gpt-image-2)")
    parser.add_argument("--num-images", type=int, default=1, choices=[1, 2, 3, 4])
    parser.add_argument("--aspect-ratio", default="1:1",
                        choices=["auto", "21:9", "16:9", "3:2", "4:3", "5:4", "1:1", "4:5", "3:4", "2:3", "9:16"],
                        help="Used by gemini-3-pro directly; mapped to a preset for gpt-image-2")
    parser.add_argument("--output-format", default="png", choices=["jpeg", "png", "webp"])
    # Gemini-only
    parser.add_argument("--resolution", default="1K", choices=["1K", "2K", "4K"],
                        help="Gemini 3 Pro only")
    parser.add_argument("--safety-tolerance", default="4", choices=["1", "2", "3", "4", "5", "6"],
                        help="Gemini 3 Pro (fal.ai) only")
    parser.add_argument("--seed", type=int, default=None,
                        help="Gemini 3 Pro (fal.ai) only")
    # GPT-Image-2 only
    parser.add_argument("--quality", default="high", choices=["low", "medium", "high"],
                        help="GPT-Image-2 only (affects cost)")
    parser.add_argument("--image-size", default=None,
                        help="GPT-Image-2 only. Preset name (e.g. landscape_4_3) or WxH (e.g. 1024x1024). Overrides --aspect-ratio mapping.")
    parser.add_argument("--image-urls", nargs="+", default=None,
                        help="GPT-Image-2 edit endpoint: one or more reference image URLs")
    parser.add_argument("--mask-url", default=None,
                        help="GPT-Image-2 edit endpoint: optional mask URL")

    parser.add_argument("--output-dir", default=".")
    parser.add_argument("--filename-prefix", default="illustration")
    parser.add_argument("--backend", default=None, choices=["fal", "openrouter"],
                        help="Force a specific backend (default: auto-detect from .env keys)")
    args = parser.parse_args()

    keys = load_env_keys()
    backend = args.backend or detect_backend(keys)

    # GPT-Image-2 is only available via fal.ai.
    if args.model == "gpt-image-2":
        if backend == "openrouter":
            if "FAL_KEY" not in keys:
                print("ERROR: --model gpt-image-2 requires FAL_KEY (not available on OpenRouter).", file=sys.stderr)
                sys.exit(1)
            print("[router] Note: gpt-image-2 is fal.ai-only — switching backend to fal.", file=sys.stderr)
            backend = "fal"
        elif backend == "none":
            print("ERROR: --model gpt-image-2 requires FAL_KEY in .env. Get one at https://fal.ai/dashboard/keys", file=sys.stderr)
            sys.exit(1)

    if backend == "none":
        print("ERROR: No API key found in .env file.", file=sys.stderr)
        print("Please add one of the following to the .env file in the skill directory:", file=sys.stderr)
        print("  OPENROUTER_API_KEY=your_openrouter_key   (https://openrouter.ai/keys)", file=sys.stderr)
        print("  FAL_KEY=your_fal_key                     (https://fal.ai/dashboard/keys)", file=sys.stderr)
        sys.exit(1)

    print(f"[router] model={args.model} backend={backend}", file=sys.stderr)

    if args.model == "gpt-image-2":
        result = generate_via_fal_gpt_image_2(
            api_key=keys["FAL_KEY"],
            prompt=args.prompt,
            num_images=args.num_images,
            aspect_ratio=args.aspect_ratio,
            image_size=args.image_size,
            quality=args.quality,
            output_format=args.output_format,
            image_urls=args.image_urls,
            mask_url=args.mask_url,
            output_dir=args.output_dir,
            filename_prefix=args.filename_prefix,
        )
    elif backend == "openrouter":
        result = generate_via_openrouter(
            api_key=keys["OPENROUTER_API_KEY"],
            prompt=args.prompt,
            num_images=args.num_images,
            aspect_ratio=args.aspect_ratio,
            resolution=args.resolution,
            output_format=args.output_format,
            output_dir=args.output_dir,
            filename_prefix=args.filename_prefix,
        )
    elif backend == "fal":
        result = generate_via_fal_gemini(
            api_key=keys["FAL_KEY"],
            prompt=args.prompt,
            num_images=args.num_images,
            aspect_ratio=args.aspect_ratio,
            output_format=args.output_format,
            resolution=args.resolution,
            safety_tolerance=args.safety_tolerance,
            seed=args.seed,
            output_dir=args.output_dir,
            filename_prefix=args.filename_prefix,
        )
    else:
        print(f"ERROR: Unknown backend '{backend}'", file=sys.stderr)
        sys.exit(1)

    print(json.dumps(result, indent=2))


if __name__ == "__main__":
    main()

File: contextual-illustrator/skills/contextual-illustrator/scripts/generate_image.mjs
#!/usr/bin/env node
// Generate images via fal.ai or OpenRouter.
// Same flags + JSON output as scripts/generate_image.py.
// Runs on Node >= 18 (global fetch) and Bun. Zero npm deps.

import { existsSync, mkdirSync, readFileSync, writeFileSync } from 'node:fs';
import { dirname, join } from 'node:path';
import { fileURLToPath } from 'node:url';
import { argv, cwd, env, exit, stderr, stdout } from 'node:process';
import { Buffer } from 'node:buffer';

const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);

// ---------------------------------------------------------------------------
// .env loading (skill root first, then walk up from cwd)
// ---------------------------------------------------------------------------

function findEnvFile() {
  const skillRoot = dirname(__dirname);
  const skillEnv = join(skillRoot, '.env');
  if (existsSync(skillEnv)) return skillEnv;

  let p = cwd();
  while (true) {
    const candidate = join(p, '.env');
    if (existsSync(candidate)) return candidate;
    const parent = dirname(p);
    if (parent === p) return null;
    p = parent;
  }
}

function loadEnvKeys() {
  const keys = {};
  for (const k of ['FAL_KEY', 'OPENROUTER_API_KEY']) {
    if (env[k]) keys[k] = env[k];
  }
  const envPath = findEnvFile();
  if (!envPath) return keys;
  const text = readFileSync(envPath, 'utf8');
  for (let line of text.split(/\r?\n/)) {
    line = line.trim();
    if (!line || line.startsWith('#')) continue;
    const eq = line.indexOf('=');
    if (eq < 0) continue;
    const k = line.slice(0, eq).trim();
    let v = line.slice(eq + 1).trim();
    if ((v.startsWith('"') && v.endsWith('"')) || (v.startsWith("'") && v.endsWith("'"))) {
      v = v.slice(1, -1);
    }
    if ((k === 'FAL_KEY' || k === 'OPENROUTER_API_KEY') && v) keys[k] = v;
  }
  return keys;
}

function detectBackend(keys) {
  if (keys.FAL_KEY) return 'fal';
  if (keys.OPENROUTER_API_KEY) return 'openrouter';
  return 'none';
}

// ---------------------------------------------------------------------------
// argv parsing
// ---------------------------------------------------------------------------

function die(msg) {
  stderr.write(`ERROR: ${msg}\n`);
  exit(1);
}

function printHelp() {
  stderr.write(`Usage: generate_image.mjs <prompt> [flags]

Same flags as scripts/generate_image.py:
  --model {gemini-3-pro|gpt-image-2}    default gpt-image-2
  --num-images {1|2|3|4}                default 1
  --aspect-ratio {auto|21:9|16:9|3:2|4:3|5:4|1:1|4:5|3:4|2:3|9:16}
  --output-format {jpeg|png|webp}        default png
  --resolution {1K|2K|4K}                Gemini only, default 1K
  --safety-tolerance {1..6}              Gemini fal only, default 4
  --seed <int>                           Gemini fal only
  --quality {low|medium|high}            GPT-Image-2 only, default high
  --image-size <preset|WxH>              GPT-Image-2 only
  --image-urls <url> [<url> ...]         GPT-Image-2 edit
  --mask-url <url>                       GPT-Image-2 edit
  --output-dir <path>                    default .
  --filename-prefix <name>               default illustration
  --backend {fal|openrouter}             default auto-detect from .env

Run with:  node scripts/generate_image.mjs <args>   (Node >= 18)
       or  bun  scripts/generate_image.mjs <args>
`);
}

const ENUMS = {
  model: ['gemini-3-pro', 'gpt-image-2'],
  numImages: [1, 2, 3, 4],
  aspectRatio: ['auto', '21:9', '16:9', '3:2', '4:3', '5:4', '1:1', '4:5', '3:4', '2:3', '9:16'],
  outputFormat: ['jpeg', 'png', 'webp'],
  resolution: ['1K', '2K', '4K'],
  safetyTolerance: ['1', '2', '3', '4', '5', '6'],
  quality: ['low', 'medium', 'high'],
  backend: ['fal', 'openrouter'],
};

function parseArgs(args) {
  const out = {
    prompt: null,
    model: 'gpt-image-2',
    numImages: 1,
    aspectRatio: '1:1',
    outputFormat: 'png',
    resolution: '1K',
    safetyTolerance: '4',
    seed: null,
    quality: 'high',
    imageSize: null,
    imageUrls: null,
    maskUrl: null,
    outputDir: '.',
    filenamePrefix: 'illustration',
    backend: null,
  };

  const requireEnum = (name, val, list) => {
    if (!list.includes(val)) die(`--${name} must be one of: ${list.join(', ')}`);
  };

  let i = 0;
  while (i < args.length) {
    const a = args[i];
    if (a === '-h' || a === '--help') {
      printHelp();
      exit(0);
    } else if (!a.startsWith('-')) {
      if (out.prompt === null) out.prompt = a;
      else die(`unexpected positional argument: ${a}`);
      i++;
    } else if (a === '--model') {
      out.model = args[++i]; requireEnum('model', out.model, ENUMS.model); i++;
    } else if (a === '--num-images') {
      out.numImages = parseInt(args[++i], 10);
      if (!ENUMS.numImages.includes(out.numImages)) die('--num-images must be 1-4');
      i++;
    } else if (a === '--aspect-ratio') {
      out.aspectRatio = args[++i]; requireEnum('aspect-ratio', out.aspectRatio, ENUMS.aspectRatio); i++;
    } else if (a === '--output-format') {
      out.outputFormat = args[++i]; requireEnum('output-format', out.outputFormat, ENUMS.outputFormat); i++;
    } else if (a === '--resolution') {
      out.resolution = args[++i]; requireEnum('resolution', out.resolution, ENUMS.resolution); i++;
    } else if (a === '--safety-tolerance') {
      out.safetyTolerance = args[++i]; requireEnum('safety-tolerance', out.safetyTolerance, ENUMS.safetyTolerance); i++;
    } else if (a === '--seed') {
      out.seed = parseInt(args[++i], 10);
      if (Number.isNaN(out.seed)) die('--seed must be an integer');
      i++;
    } else if (a === '--quality') {
      out.quality = args[++i]; requireEnum('quality', out.quality, ENUMS.quality); i++;
    } else if (a === '--image-size') {
      out.imageSize = args[++i]; i++;
    } else if (a === '--image-urls') {
      out.imageUrls = [];
      i++;
      while (i < args.length && !args[i].startsWith('-')) {
        out.imageUrls.push(args[i++]);
      }
    } else if (a === '--mask-url') {
      out.maskUrl = args[++i]; i++;
    } else if (a === '--output-dir') {
      out.outputDir = args[++i]; i++;
    } else if (a === '--filename-prefix') {
      out.filenamePrefix = args[++i]; i++;
    } else if (a === '--backend') {
      out.backend = args[++i]; requireEnum('backend', out.backend, ENUMS.backend); i++;
    } else {
      die(`unknown flag: ${a}`);
    }
  }

  if (out.prompt === null) {
    printHelp();
    die('prompt is required');
  }
  return out;
}

// ---------------------------------------------------------------------------
// shared helpers
// ---------------------------------------------------------------------------

const GPT2_ASPECT_TO_SIZE = {
  '1:1': 'square_hd',
  '16:9': 'landscape_16_9',
  '4:3': 'landscape_4_3',
  '3:2': 'landscape_4_3',
  '5:4': 'square_hd',
  '4:5': 'portrait_4_3',
  '3:4': 'portrait_4_3',
  '2:3': 'portrait_4_3',
  '9:16': 'portrait_16_9',
  '21:9': 'landscape_16_9',
  'auto': 'landscape_4_3',
};

function resolveGpt2ImageSize(imageSize, aspectRatio) {
  if (imageSize) {
    if (imageSize.toLowerCase().includes('x')) {
      const [w, h] = imageSize.toLowerCase().split('x', 2);
      const wn = parseInt(w, 10);
      const hn = parseInt(h, 10);
      if (Number.isNaN(wn) || Number.isNaN(hn)) {
        die(`invalid --image-size '${imageSize}', expected WxH or preset name`);
      }
      return { width: wn, height: hn };
    }
    return imageSize;
  }
  return GPT2_ASPECT_TO_SIZE[aspectRatio] || 'landscape_4_3';
}

function ensureDir(d) {
  if (!existsSync(d)) mkdirSync(d, { recursive: true });
}

async function downloadOrDecode(url, filepath) {
  if (url.startsWith('data:')) {
    const comma = url.indexOf(',');
    if (comma < 0) throw new Error('malformed data URL');
    const b64 = url.slice(comma + 1);
    writeFileSync(filepath, Buffer.from(b64, 'base64'));
    return false; // not a remote URL
  }
  const resp = await fetch(url);
  if (!resp.ok) throw new Error(`download ${url} -> HTTP ${resp.status}`);
  writeFileSync(filepath, Buffer.from(await resp.arrayBuffer()));
  return true;
}

// ---------------------------------------------------------------------------
// fal.ai — sync mode (one POST, final JSON back)
// ---------------------------------------------------------------------------

async function callFalSync(endpoint, apiKey, body) {
  const url = `https://fal.run/${endpoint}`;
  const resp = await fetch(url, {
    method: 'POST',
    headers: {
      'Authorization': `Key ${apiKey}`,
      'Content-Type': 'application/json',
    },
    body: JSON.stringify(body),
  });
  if (!resp.ok) {
    const text = await resp.text();
    die(`fal.ai returned HTTP ${resp.status}: ${text.slice(0, 800)}`);
  }
  return await resp.json();
}

async function generateViaFalGpt2(opts, apiKey) {
  const isEdit = Array.isArray(opts.imageUrls) && opts.imageUrls.length > 0;
  const endpoint = isEdit ? 'openai/gpt-image-2/edit' : 'openai/gpt-image-2';

  const body = {
    prompt: opts.prompt,
    num_images: opts.numImages,
    quality: opts.quality,
    output_format: opts.outputFormat,
  };

  if (isEdit) {
    body.image_urls = opts.imageUrls;
    if (opts.maskUrl) body.mask_url = opts.maskUrl;
    if (opts.imageSize) body.image_size = resolveGpt2ImageSize(opts.imageSize, opts.aspectRatio);
  } else {
    body.image_size = resolveGpt2ImageSize(opts.imageSize, opts.aspectRatio);
  }

  const tag = `fal:${endpoint}`;
  stderr.write(`[${tag}] Sending request...\n`);
  const result = await callFalSync(endpoint, apiKey, body);

  ensureDir(opts.outputDir);
  const files = [];
  const urls = [];
  const images = result.images ?? [];
  for (let i = 0; i < images.length; i++) {
    const u = images[i].url;
    const suffix = images.length > 1 ? `_${i + 1}` : '';
    const filename = `${opts.filenamePrefix}${suffix}.${opts.outputFormat}`;
    const filepath = join(opts.outputDir, filename);
    const wasRemote = await downloadOrDecode(u, filepath);
    if (wasRemote) urls.push(u);
    files.push(filepath);
    stderr.write(`[${tag}] Saved: ${filepath}\n`);
  }

  return {
    backend: 'fal',
    model: 'gpt-image-2',
    endpoint,
    files,
    urls,
    description: result.description ?? '',
  };
}

async function generateViaFalGemini(opts, apiKey) {
  const endpoint = 'fal-ai/gemini-3-pro-image-preview';
  const body = {
    prompt: opts.prompt,
    num_images: opts.numImages,
    aspect_ratio: opts.aspectRatio,
    output_format: opts.outputFormat,
    safety_tolerance: opts.safetyTolerance,
    resolution: opts.resolution,
  };
  if (opts.seed !== null) body.seed = opts.seed;

  const tag = `fal:${endpoint}`;
  stderr.write(`[${tag}] Sending request...\n`);
  const result = await callFalSync(endpoint, apiKey, body);

  ensureDir(opts.outputDir);
  const files = [];
  const urls = [];
  const images = result.images ?? [];
  for (let i = 0; i < images.length; i++) {
    const u = images[i].url;
    const suffix = images.length > 1 ? `_${i + 1}` : '';
    const filename = `${opts.filenamePrefix}${suffix}.${opts.outputFormat}`;
    const filepath = join(opts.outputDir, filename);
    const wasRemote = await downloadOrDecode(u, filepath);
    if (wasRemote) urls.push(u);
    files.push(filepath);
    stderr.write(`[${tag}] Saved: ${filepath}\n`);
  }

  return {
    backend: 'fal',
    model: 'gemini-3-pro',
    files,
    urls,
    description: result.description ?? '',
  };
}

// ---------------------------------------------------------------------------
// OpenRouter (Gemini 3 Pro only)
// ---------------------------------------------------------------------------

async function generateViaOpenRouter(opts, apiKey) {
  const url = 'https://openrouter.ai/api/v1/chat/completions';
  let promptStr = opts.prompt;
  if (opts.numImages > 1) promptStr += ` Generate ${opts.numImages} different variations.`;

  const imageConfig = {};
  if (opts.aspectRatio && opts.aspectRatio !== 'auto') imageConfig.aspect_ratio = opts.aspectRatio;
  if (opts.resolution) imageConfig.image_size = opts.resolution;

  const body = {
    model: 'google/gemini-3-pro-image-preview',
    messages: [{ role: 'user', content: promptStr }],
    modalities: ['image', 'text'],
  };
  if (Object.keys(imageConfig).length) body.image_config = imageConfig;

  stderr.write(`[openrouter] Sending request...\n`);
  const resp = await fetch(url, {
    method: 'POST',
    headers: {
      'Authorization': `Bearer ${apiKey}`,
      'Content-Type': 'application/json',
    },
    body: JSON.stringify(body),
  });
  if (!resp.ok) {
    const text = await resp.text();
    die(`OpenRouter HTTP ${resp.status}: ${text.slice(0, 800)}`);
  }
  const result = await resp.json();
  const message = result.choices?.[0]?.message ?? {};

  // Canonical: message.images[].image_url.url. Defensive fallback: content array parts.
  const imageUrls = [];
  if (Array.isArray(message.images)) {
    for (const im of message.images) {
      const u = im?.image_url?.url;
      if (u) imageUrls.push(u);
    }
  }
  if (imageUrls.length === 0 && Array.isArray(message.content)) {
    for (const part of message.content) {
      if (part?.type === 'image_url' && part?.image_url?.url) imageUrls.push(part.image_url.url);
    }
  }

  ensureDir(opts.outputDir);
  const files = [];
  const urls = [];
  for (let i = 0; i < imageUrls.length; i++) {
    const u = imageUrls[i];
    const suffix = imageUrls.length > 1 ? `_${i + 1}` : '';
    const filename = `${opts.filenamePrefix}${suffix}.${opts.outputFormat}`;
    const filepath = join(opts.outputDir, filename);
    const wasRemote = await downloadOrDecode(u, filepath);
    if (wasRemote) urls.push(u);
    files.push(filepath);
    stderr.write(`[openrouter] Saved: ${filepath}\n`);
  }

  let description = '';
  if (typeof message.content === 'string') description = message.content;
  else if (Array.isArray(message.content)) {
    description = message.content
      .filter((p) => p?.type === 'text')
      .map((p) => p?.text ?? '')
      .join('\n');
  }

  return {
    backend: 'openrouter',
    model: 'gemini-3-pro',
    files,
    urls,
    description,
  };
}

// ---------------------------------------------------------------------------
// main
// ---------------------------------------------------------------------------

async function main() {
  const opts = parseArgs(argv.slice(2));
  const keys = loadEnvKeys();
  let backend = opts.backend ?? detectBackend(keys);

  if (opts.model === 'gpt-image-2') {
    if (backend === 'openrouter') {
      if (!keys.FAL_KEY) die('--model gpt-image-2 requires FAL_KEY (not available on OpenRouter).');
      stderr.write('[router] Note: gpt-image-2 is fal.ai-only — switching backend to fal.\n');
      backend = 'fal';
    } else if (backend === 'none') {
      die('--model gpt-image-2 requires FAL_KEY in .env. Get one at https://fal.ai/dashboard/keys');
    }
  }

  if (backend === 'none') {
    stderr.write('ERROR: No API key found in .env file.\n');
    stderr.write('Please add one of the following to the .env file in the skill directory:\n');
    stderr.write('  OPENROUTER_API_KEY=your_openrouter_key   (https://openrouter.ai/keys)\n');
    stderr.write('  FAL_KEY=your_fal_key                     (https://fal.ai/dashboard/keys)\n');
    exit(1);
  }

  stderr.write(`[router] model=${opts.model} backend=${backend}\n`);

  let result;
  if (opts.model === 'gpt-image-2') {
    result = await generateViaFalGpt2(opts, keys.FAL_KEY);
  } else if (backend === 'openrouter') {
    result = await generateViaOpenRouter(opts, keys.OPENROUTER_API_KEY);
  } else if (backend === 'fal') {
    result = await generateViaFalGemini(opts, keys.FAL_KEY);
  } else {
    die(`Unknown backend '${backend}'`);
  }

  stdout.write(JSON.stringify(result, null, 2) + '\n');
}

main().catch((err) => {
  stderr.write(`ERROR: ${err.message ?? err}\n`);
  exit(1);
});
